### DISCUSSION TITLE
Support external files/links as attachments

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5605] [Feature Requests]
Based on the recent meeting [Open Discussion on Associated Media Management in Specify 7](https://discourse.specifysoftware.org/t/open-discussion-on-associated-media-management-in-specify-7/2156/1) (took place on December 12, 2024).

A significant point of consensus emerged regarding the need for the capability to attach external URLs to Specify records. This feature was met with overwhelming support, as 17 participants expressed their approval through thumbs-up reactions in the chat among other spoken responses in our meeting.

## Support for External URLs as attachments

The idea of allowing external URLs to be used as attachments resonated strongly with the community. The following individuals voiced their support for this functionality:

-  **Warren Brown** (University of Florida) _Founding Partner_
-  **Robert Cubey** (Royal Botanic Gardens Edinburgh)
-  **Corinne Fuchs** (Florida Fish and Wildlife Conservation Commission)
-  **Cody Thompson** (University of Michigan)
-  **Matthew Cruz** (University of Michigan)
-  **Natalia Lopez Carranza** (University of Kansas)
-  **Kendra Driscoll** (New Brunswick Museum)
-  **Tamaki Yuri** (Ohio State University)
-  **Miquel Martinez** (Museu de Ciències Naturals de Barcelona)
-  **Nate Shoobs** (Ohio State University)
-  **Dianna Krejsa** (University of Kansas)
-  **Andy Bentley** (University of Kansas)
-  **Faith Zimmerman** (University of Michigan)
-  **Heather C** (Agriculture and Agri-Food Canada)
-  **Mark Pitblado** (University of British Columbia)
-  **Ashley Ferguson** (College of Idaho)
-  **Zach Randall** 

This collective endorsement underscores the importance of integrating external URLs into the Specify system, allowing users to link to relevant resources, thereby enhancing the functionality and usability of the software.

## Use Cases for External URLs as attachments

Participants shared various use cases for integrating external URLs into their workflows:

1. **Linking to High-Resolution Images**: 
   - **Matthew Cruz** highlighted the need for an "External Asset" type that would allow users to maintain low-resolution images for fast rendering while providing links to high-resolution assets stored externally. This would streamline the management of digital assets and improve the efficiency of data retrieval.

2. **Accessing 3D Models**: 
   - **Cody Thompson** mentioned using links to 3D scans and models hosted on external platforms like Sketchfab and MorphoSource. The ability to link these resources directly within Specify would facilitate easier access to valuable visual data that supports research and education.

3. **Compliance with External Databases**: 
   - **Warren Brown** pointed out the necessity of linking to external repositories for compliance with legal and ethical standards, stating that unique URLs for each asset would enable better management of digital assets in accordance with institutional policies.

4. **Enhanced Data Sharing**: 
   - **Corinne Fuchs** expressed interest in tying specimens to social media posts, which could help itemize the use of the collection in outreach efforts. This capability would enhance public engagement and visibility for collections.

5. **Streamlining Workflow**: 
   - **Nate Shoobs** noted that integrating external URLs would simplify workflows by reducing the need to duplicate media files within Specify, allowing users to link directly to existing resources instead.

## Example Interaction

Users would like the ability add URLs as an attachment via Specify:

<img width="400" alt="Screenshot 2024-12-13 at 12 16 57 PM" src="https://github.com/user-attachments/assets/ac45eb00-f848-46c2-9957-223a04c0d144" />

Note that you would need a "next" button after entering the URL or some mechanism to capture metadata about it.

### **Draft Requirements:**

* Users should be able to add a URL as an attachment to a record via the forms interface along with associated metadata
* Users should be able to import attachment URLs via the WorkBench by mapping one URL per column to the Attachment table
* Users should be able to query and return a list of attachment URLs, both for assets within the Specify Asset Server and those that exist externally
* Users should be able to aggregate and export these URLs as a single field to align with the [dwc:associatedMedia](https://dwc.tdwg.org/list/#dwc_associatedMedia) term
* Users should be able to build a separate query to return these URLs individually with metadata for use with the [ac:accessURI](http://rs.tdwg.org/ac/terms/accessURI) term

### Requested Changes

* Addition of a `resourceURL` field to the `Attachment` table
* New interface for the user to enter the URL
* Updated table format/aggregation to accomodate the new field
* New capability to automatically construct the image URL for assets hosted locally based on the asset server base URL to avoid the need to configure one manually (see this [guide for information on how users currently workaround this](https://discourse.specifysoftware.org/t/show-attachment-urls-in-queries-and-when-data-exporting/1212))

## Conclusion

The feedback from participants clearly indicates a strong demand for the ability to attach external URLs as part of the asset management process within Specify. By implementing this feature, Specify can enhance its functionality, improve user experience, and better support the diverse needs of its community.

---
_Original issue body:_

Attachments should be more versatile. It should include URLs, links to images, documents, etc. 

**Requested by:** Matthew Cruz & others

**Comment by maxpatiiuk** [comment]
Until this feature is added, they can modify the asset server to get data from a different source.

For example, here is a modification that stores/fetches attachments in an s3 bucket on Digital Ocean - https://github.com/specify/web-asset-server/tree/development

**Comment by grantfitzsimmons** [comment]
This solution would not enable storing attachments on the server in addition to external servers right?

**Comment by maxpatiiuk** [comment]
Yeah, you are correct, not a good workaround.

Implementing this feature won't be hard. Probably the hardest part would be figuring out how to handle attachment thumbnails for external attachments

**Comment by grantfitzsimmons** [comment]
Worst case- use the file extension generate a thumbnail that corresponds with the file type (blank file with PNG, etc)

Best case- use something like [embed to grab the thumbnail.](https://github.com/oscarotero/Embed)

**Comment by grantfitzsimmons** [comment]
Would the burden not be on the browser accessing 7 to load the image rather than the DB's server?

**Comment by mc0822** [comment]
Ben Ritchie of RBGE requested this feature also, on 12/2/22.

**Comment by benritchie** [comment]
just to expand on this - our use case is around having an existing Image server, and wanting to embed images from there.  
1 key point is - its not going to be feasible to add every image individually as attachments to the Collection objects.  I'd be keen to see a query model (i.e. Specify looks up the Barcode in the external system, and requests all the images - and then displays them) rather than a model where Specify knows the list of images to retrieve.  
- so I think this is meaningfully different from the existing Attachements model - where image existence is known by specify.  
Thanks

**Comment by grantfitzsimmons** [comment]
> This is a big change, but other systems support external linkages as attachments (including thumbnails). Attachments stored in MorphoSource or other destinations currently cannot be linked traditionally from within Specify since all attachments must exist on the asset server.

**Comment by grantfitzsimmons** [comment]
> Like weblinks, but using the current attachment record viewer architecture. 
> 
> > Scenario
> > A lot of our collection objects have related information (such as images, DNA sequences) that are held in other CSIRO systems. To close the loop, we are planning to create a link between the systems to exchange basic information via APIs. For our Specify instances that means we are planning to pull back urls that point to the object/information in the other system. For example, we may have an object that has three images held in CSIRO’s Data Access Portal (DAP) and the plan is to display these links ‘somewhere’ in the object’s record as a weblink so the user has the ability to jump straight over to the other system to view the image.
> > 
> > Feature request
> > We were thinking that at least for the images it’d be great if these links could be stored in the attachment server with an ability to pull back and display a thumbnail of the image.
> > 
From @realVinayak on https://github.com/specify/specify7/issues/4485

**Comment by acbentley** [comment]
The attachment URL option should also invoke the attachment table form that would allow the addition of metadata associated with the attachment while also showing a thumbnail if possible.  Additionally, a regular attachment should also populate the URL field with the concatenated attachment server address enabling this field to be mapped the DwC extension field for all attachments.

**Comment by specifysoftware** [comment]
This issue has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/open-discussion-on-associated-media-management-in-specify-7/2156/3
